<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="This package is for the eSVD-DE, a way to test for differential expression for single-cell RNA-seq data when many donors/individuals are sequenced, and you wish to test for differential expression at the individual-level (i.e., among individuals), not at cell-level. Specifically, this method uses the eSVD matrix factorization to pool information among genes together, and then carefully adjusts the denoised gene expression to not inflate the Type-1 error. The eSVD matrix factorization (which is the backbone of the eSVD-DE method) assumes that each row and column of a matrix is associated with a low-dimensional vector, and eSVD estimates an embedding for each cell with respect to a hierarchical model where the inner product between the rows and columns latent vectors is the natural parameter of a one-parameter exponential family random variable.">
<title>eSVD2 for performing the eSVD-DE for cohort-wide differential expression analysis • eSVD2</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="eSVD2 for performing the eSVD-DE for cohort-wide differential expression analysis">
<meta property="og:description" content="This package is for the eSVD-DE, a way to test for differential expression for single-cell RNA-seq data when many donors/individuals are sequenced, and you wish to test for differential expression at the individual-level (i.e., among individuals), not at cell-level. Specifically, this method uses the eSVD matrix factorization to pool information among genes together, and then carefully adjusts the denoised gene expression to not inflate the Type-1 error. The eSVD matrix factorization (which is the backbone of the eSVD-DE method) assumes that each row and column of a matrix is associated with a low-dimensional vector, and eSVD estimates an embedding for each cell with respect to a hierarchical model where the inner product between the rows and columns latent vectors is the natural parameter of a one-parameter exponential family random variable.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">eSVD2</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="articles/eSVD2.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="articles/asd-preprocess.html">Neuron preprocessing for eSVD-DE tutorial</a>
    <a class="dropdown-item" href="articles/asd.html">Neuron analysis for Layer 2/3 in ASD</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="purpose">Purpose<a class="anchor" aria-label="anchor" href="#purpose"></a>
</h1></div>
<p>This repository contains all the functions to perform eSVD-DE (for package <code>eSVD2</code>, version <code>1.0.0</code> as of June 16, 2024) and the downstream analysis, for the paper “eSVD-DE: cohort-wide differential expression in single-cell RNA-seq data using exponential-family embeddings”. See the companion GitHub package <a href="https://github.com/linnykos/eSVD2_examples" class="external-link uri">https://github.com/linnykos/eSVD2_examples</a> for all the analyses performed in the paper. (Note, the original analysis was performed on eSVD2 version 0.0.0.0071, and we are working to make the code in <a href="https://github.com/linnykos/eSVD2_examples" class="external-link uri">https://github.com/linnykos/eSVD2_examples</a> to be compatibile with the latest version of eSVD2.)</p>
<p>This code was developed and tested primarily on R 4.3.2. on a 2023 Macbook Pro (macOS Sonoma 14.2.1) equipped with Apple M2 Max processor (32 Gb RAM).</p>
<!-- badges: start -->

</div>
<div class="section level1">
<h1 id="tutorials-and-vignettes">Tutorials and vignettes<a class="anchor" aria-label="anchor" href="#tutorials-and-vignettes"></a>
</h1>
<p>Please see <a href="https://linnykos.github.io/eSVD2/index.html" class="uri">https://linnykos.github.io/eSVD2/index.html</a> for in-depth tutorials and vignettes on how to use Tilted-CCA.</p>
</div>
<div class="section level1">
<h1 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h1>
<p>This package can be installed through <code>devtools</code> in R.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://devtools.r-lib.org/" class="external-link">"devtools"</a></span><span class="op">)</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"linnykos/eSVD2"</span><span class="op">)</span></span></code></pre></div>
<p>The package itself depends on several packages. These include <code>irlba</code>, <code>glmnet</code>, <code>locfdr</code>, <code>Matrix</code>, <code>matrixStats</code>, <code>Rmpfr</code>, <code>RSpectra</code>, and <code>sparseMatrixStats</code>. See the last section of this README to see where (i.e., CRAN, Bioconductor, or GitHub) to download all such packages. We have noted that <code>Rmpfr</code> is sometimes tricky to install due to its required C++ libraries.</p>
<p>After installation of all the dependencies, the installation of the <code>tiltedCCA</code> package itself takes modest time (less than 10 minutes). The installation time mainly consists of time to compile the C++ code since the matrix factorization optimization was written using Rcpp for faster performance.</p>
<details><summary><strong>Known installation issues and the solutions</strong>
</summary><p>(Solution posted on December 24, 2023): If you come across the error,</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a>Error <span class="cf">in</span> irlba<span class="sc">::</span><span class="fu">irlba</span>() <span class="sc">:</span> </span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a>  <span class="cf">function</span> <span class="st">'as_cholmod_sparse'</span> not provided by package <span class="st">'Matrix'</span></span></code></pre></div>
<p>then it is likely you need to downgrade your version of <code>Matrix</code> to <code>1.6-1.1</code>. See <a href="https://github.com/bwlewis/irlba/issues/70#issuecomment-1826900769" class="external-link uri">https://github.com/bwlewis/irlba/issues/70#issuecomment-1826900769</a>. Hence, in the R console,</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/remove.packages.html" class="external-link">remove.packages</a></span><span class="op">(</span><span class="st">"SeuratObject"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/remove.packages.html" class="external-link">remove.packages</a></span><span class="op">(</span><span class="st">"Matrix"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/remove.packages.html" class="external-link">remove.packages</a></span><span class="op">(</span><span class="st">"irlba"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"Matrix"</span>, type <span class="op">=</span> <span class="st">"source"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"irlba"</span>, type <span class="op">=</span> <span class="st">"source"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"SeuratObject"</span><span class="op">)</span></span></code></pre></div>
</details>
</div>
<div class="section level1">
<h1 id="small-simulated-dataset-to-demo-the-software">Small simulated dataset to demo the software<a class="anchor" aria-label="anchor" href="#small-simulated-dataset-to-demo-the-software"></a>
</h1>
<p>See <a href="https://linnykos.github.io/eSVD2/articles/eSVD2.html" class="uri">https://linnykos.github.io/eSVD2/articles/eSVD2.html</a> for the small demo on how to use eSVD2.</p>
</div>
<div class="section level1">
<h1 id="setup">Setup<a class="anchor" aria-label="anchor" href="#setup"></a>
</h1>
<p>The following shows the suggested package versions that the developer (GitHub username: linnykos) used when developing the eSVD2 package.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="sc">&gt;</span> devtools<span class="sc">::</span><span class="fu">session_info</span>()</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a>─ Session info ──────────────────────────</span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a> setting  value</span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a> version  R version <span class="dv">4</span>.<span class="fl">3.2</span> (<span class="dv">2023-10-31</span>)</span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a> os       macOS Sonoma <span class="dv">14</span>.<span class="fl">2.1</span></span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a> system   aarch64, darwin20</span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a> ui       RStudio</span>
<span id="cb4-8"><a href="#cb4-8" tabindex="-1"></a> <span class="fu">language</span> (EN)</span>
<span id="cb4-9"><a href="#cb4-9" tabindex="-1"></a> collate  en_US.UTF<span class="dv">-8</span></span>
<span id="cb4-10"><a href="#cb4-10" tabindex="-1"></a> ctype    en_US.UTF<span class="dv">-8</span></span>
<span id="cb4-11"><a href="#cb4-11" tabindex="-1"></a> tz       America<span class="sc">/</span>Chicago</span>
<span id="cb4-12"><a href="#cb4-12" tabindex="-1"></a> date     <span class="dv">2024-06-16</span></span>
<span id="cb4-13"><a href="#cb4-13" tabindex="-1"></a> rstudio  <span class="dv">2024</span>.<span class="fl">04.1</span><span class="sc">+</span><span class="dv">748</span> Chocolate <span class="fu">Cosmos</span> (desktop)</span>
<span id="cb4-14"><a href="#cb4-14" tabindex="-1"></a> pandoc   <span class="cn">NA</span></span>
<span id="cb4-15"><a href="#cb4-15" tabindex="-1"></a></span>
<span id="cb4-16"><a href="#cb4-16" tabindex="-1"></a>─ Packages ──────────────────────────────</span>
<span id="cb4-17"><a href="#cb4-17" tabindex="-1"></a> package     <span class="sc">*</span> version <span class="fu">date</span> (UTC) lib source</span>
<span id="cb4-18"><a href="#cb4-18" tabindex="-1"></a> brio          <span class="dv">1</span>.<span class="fl">1.5</span>   <span class="dv">2024-04-24</span> [<span class="dv">1</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">3.1</span>)</span>
<span id="cb4-19"><a href="#cb4-19" tabindex="-1"></a> cachem        <span class="dv">1</span>.<span class="fl">1.0</span>   <span class="dv">2024-05-16</span> [<span class="dv">1</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">3.3</span>)</span>
<span id="cb4-20"><a href="#cb4-20" tabindex="-1"></a> cli           <span class="dv">3</span>.<span class="fl">6.2</span>   <span class="dv">2023-12-11</span> [<span class="dv">1</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">3.1</span>)</span>
<span id="cb4-21"><a href="#cb4-21" tabindex="-1"></a> devtools      <span class="dv">2</span>.<span class="fl">4.5</span>   <span class="dv">2022-10-11</span> [<span class="dv">1</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">3.0</span>)</span>
<span id="cb4-22"><a href="#cb4-22" tabindex="-1"></a> digest        <span class="dv">0</span>.<span class="fl">6.35</span>  <span class="dv">2024-03-11</span> [<span class="dv">1</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">3.1</span>)</span>
<span id="cb4-23"><a href="#cb4-23" tabindex="-1"></a> ellipsis      <span class="dv">0</span>.<span class="fl">3.2</span>   <span class="dv">2021-04-29</span> [<span class="dv">1</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">3.0</span>)</span>
<span id="cb4-24"><a href="#cb4-24" tabindex="-1"></a> eSVD2       <span class="sc">*</span> <span class="dv">1</span>.<span class="fl">0.0</span>   <span class="dv">2024-06-17</span> [<span class="dv">1</span>] local</span>
<span id="cb4-25"><a href="#cb4-25" tabindex="-1"></a> fastmap       <span class="dv">1</span>.<span class="fl">2.0</span>   <span class="dv">2024-05-15</span> [<span class="dv">1</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">3.3</span>)</span>
<span id="cb4-26"><a href="#cb4-26" tabindex="-1"></a> fs            <span class="dv">1</span>.<span class="fl">6.4</span>   <span class="dv">2024-04-25</span> [<span class="dv">1</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">3.1</span>)</span>
<span id="cb4-27"><a href="#cb4-27" tabindex="-1"></a> glue          <span class="dv">1</span>.<span class="fl">7.0</span>   <span class="dv">2024-01-09</span> [<span class="dv">1</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">3.1</span>)</span>
<span id="cb4-28"><a href="#cb4-28" tabindex="-1"></a> htmltools     <span class="dv">0</span>.<span class="dv">5</span>.<span class="fl">8.1</span> <span class="dv">2024-04-04</span> [<span class="dv">1</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">3.1</span>)</span>
<span id="cb4-29"><a href="#cb4-29" tabindex="-1"></a> htmlwidgets   <span class="dv">1</span>.<span class="fl">6.4</span>   <span class="dv">2023-12-06</span> [<span class="dv">1</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">3.1</span>)</span>
<span id="cb4-30"><a href="#cb4-30" tabindex="-1"></a> httpuv        <span class="dv">1</span>.<span class="fl">6.15</span>  <span class="dv">2024-03-26</span> [<span class="dv">1</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">3.1</span>)</span>
<span id="cb4-31"><a href="#cb4-31" tabindex="-1"></a> later         <span class="dv">1</span>.<span class="fl">3.2</span>   <span class="dv">2023-12-06</span> [<span class="dv">1</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">3.1</span>)</span>
<span id="cb4-32"><a href="#cb4-32" tabindex="-1"></a> lifecycle     <span class="dv">1</span>.<span class="fl">0.4</span>   <span class="dv">2023-11-07</span> [<span class="dv">1</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">3.1</span>)</span>
<span id="cb4-33"><a href="#cb4-33" tabindex="-1"></a> magrittr      <span class="dv">2</span>.<span class="fl">0.3</span>   <span class="dv">2022-03-30</span> [<span class="dv">1</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">3.0</span>)</span>
<span id="cb4-34"><a href="#cb4-34" tabindex="-1"></a> memoise       <span class="dv">2</span>.<span class="fl">0.1</span>   <span class="dv">2021-11-26</span> [<span class="dv">1</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">3.0</span>)</span>
<span id="cb4-35"><a href="#cb4-35" tabindex="-1"></a> mime          <span class="fl">0.12</span>    <span class="dv">2021-09-28</span> [<span class="dv">1</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">3.0</span>)</span>
<span id="cb4-36"><a href="#cb4-36" tabindex="-1"></a> miniUI        <span class="dv">0</span>.<span class="dv">1</span>.<span class="fl">1.1</span> <span class="dv">2018-05-18</span> [<span class="dv">1</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">3.0</span>)</span>
<span id="cb4-37"><a href="#cb4-37" tabindex="-1"></a> pkgbuild      <span class="dv">1</span>.<span class="fl">4.4</span>   <span class="dv">2024-03-17</span> [<span class="dv">1</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">3.1</span>)</span>
<span id="cb4-38"><a href="#cb4-38" tabindex="-1"></a> pkgload       <span class="dv">1</span>.<span class="fl">3.4</span>   <span class="dv">2024-01-16</span> [<span class="dv">1</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">3.1</span>)</span>
<span id="cb4-39"><a href="#cb4-39" tabindex="-1"></a> profvis       <span class="dv">0</span>.<span class="fl">3.8</span>   <span class="dv">2023-05-02</span> [<span class="dv">1</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">3.0</span>)</span>
<span id="cb4-40"><a href="#cb4-40" tabindex="-1"></a> promises      <span class="dv">1</span>.<span class="fl">3.0</span>   <span class="dv">2024-04-05</span> [<span class="dv">1</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">3.1</span>)</span>
<span id="cb4-41"><a href="#cb4-41" tabindex="-1"></a> purrr         <span class="dv">1</span>.<span class="fl">0.2</span>   <span class="dv">2023-08-10</span> [<span class="dv">1</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">3.0</span>)</span>
<span id="cb4-42"><a href="#cb4-42" tabindex="-1"></a> R6            <span class="dv">2</span>.<span class="fl">5.1</span>   <span class="dv">2021-08-19</span> [<span class="dv">1</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">3.0</span>)</span>
<span id="cb4-43"><a href="#cb4-43" tabindex="-1"></a> Rcpp        <span class="sc">*</span> <span class="dv">1</span>.<span class="fl">0.12</span>  <span class="dv">2024-01-09</span> [<span class="dv">1</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">3.1</span>)</span>
<span id="cb4-44"><a href="#cb4-44" tabindex="-1"></a> remotes       <span class="dv">2</span>.<span class="fl">5.0</span>   <span class="dv">2024-03-17</span> [<span class="dv">1</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">3.1</span>)</span>
<span id="cb4-45"><a href="#cb4-45" tabindex="-1"></a> rlang         <span class="dv">1</span>.<span class="fl">1.4</span>   <span class="dv">2024-06-04</span> [<span class="dv">1</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">3.3</span>)</span>
<span id="cb4-46"><a href="#cb4-46" tabindex="-1"></a> rstudioapi    <span class="dv">0</span>.<span class="fl">16.0</span>  <span class="dv">2024-03-24</span> [<span class="dv">1</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">3.1</span>)</span>
<span id="cb4-47"><a href="#cb4-47" tabindex="-1"></a> sessioninfo   <span class="dv">1</span>.<span class="fl">2.2</span>   <span class="dv">2021-12-06</span> [<span class="dv">1</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">3.0</span>)</span>
<span id="cb4-48"><a href="#cb4-48" tabindex="-1"></a> shiny         <span class="dv">1</span>.<span class="dv">8</span>.<span class="fl">1.1</span> <span class="dv">2024-04-02</span> [<span class="dv">1</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">3.1</span>)</span>
<span id="cb4-49"><a href="#cb4-49" tabindex="-1"></a> stringi       <span class="dv">1</span>.<span class="fl">8.4</span>   <span class="dv">2024-05-06</span> [<span class="dv">1</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">3.1</span>)</span>
<span id="cb4-50"><a href="#cb4-50" tabindex="-1"></a> stringr       <span class="dv">1</span>.<span class="fl">5.1</span>   <span class="dv">2023-11-14</span> [<span class="dv">1</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">3.1</span>)</span>
<span id="cb4-51"><a href="#cb4-51" tabindex="-1"></a> testthat    <span class="sc">*</span> <span class="dv">3</span>.<span class="dv">2</span>.<span class="fl">1.1</span> <span class="dv">2024-04-14</span> [<span class="dv">1</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">3.1</span>)</span>
<span id="cb4-52"><a href="#cb4-52" tabindex="-1"></a> urlchecker    <span class="dv">1</span>.<span class="fl">0.1</span>   <span class="dv">2021-11-30</span> [<span class="dv">1</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">3.0</span>)</span>
<span id="cb4-53"><a href="#cb4-53" tabindex="-1"></a> usethis       <span class="dv">2</span>.<span class="fl">2.3</span>   <span class="dv">2024-02-19</span> [<span class="dv">1</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">3.1</span>)</span>
<span id="cb4-54"><a href="#cb4-54" tabindex="-1"></a> vctrs         <span class="dv">0</span>.<span class="fl">6.5</span>   <span class="dv">2023-12-01</span> [<span class="dv">1</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">3.1</span>)</span>
<span id="cb4-55"><a href="#cb4-55" tabindex="-1"></a> xtable        <span class="fl">1.8</span><span class="dv">-4</span>   <span class="dv">2019-04-21</span> [<span class="dv">1</span>] <span class="fu">CRAN</span> (R <span class="dv">4</span>.<span class="fl">3.0</span>)</span>
<span id="cb4-56"><a href="#cb4-56" tabindex="-1"></a></span>
<span id="cb4-57"><a href="#cb4-57" tabindex="-1"></a> [<span class="dv">1</span>] <span class="sc">/</span>Library<span class="sc">/</span>Frameworks<span class="sc">/</span>R.framework<span class="sc">/</span>Versions<span class="sc">/</span><span class="fl">4.3</span><span class="sc">-</span>arm64<span class="sc">/</span>Resources<span class="sc">/</span>library</span></code></pre></div>
</div>

  </main><aside class="col-md-3"><div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> +file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing eSVD2</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Kevin Z Lin <br><small class="roles"> Author, maintainer </small>  </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://doi.org/10.1186/s12859-024-05724-7" class="external-link"><img src="https://img.shields.io/badge/doi-10.1073/pnas.2303647120-firebrick.svg" alt="DOI:10.1186/s12859-024-05724-7"></a></li>
<li><a href="https://lifecycle.r-lib.org/articles/stages.html#stable" class="external-link"><img src="https://img.shields.io/badge/lifecycle-stable-brightgreen.svg" alt="Lifecycle: stable"></a></li>
</ul>
</div>

  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Kevin Z Lin.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
